### PR TITLE
ENT-1463, ENT-1903: Document the "determinisation" process.

### PR DESCRIPTION
**Pull request by chrisr3**

### CODE REVIEW

File: docs/source/deterministic-modules.rst [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+  To build a new version of this JAR and upload it to the Artifactory, see the ``create-jdk8u`` module. This is a
+  standalone Gradle project within the Corda repository that will clone the ``deterministic-jvm8`` branch of Corda's
+  `OpenJDK repository <https://github.com/corda/openjdk>`_ and then build it. (This currently requires a C++ compiler,
+  GNU Make and a UNIX-like development environment.)
+
+Corda Modules
+  ``core-deterministic`` and ``serialization-deterministic`` are generated from Corda's ``core`` and ``serialization``
+  modules respectively using both `ProGuard <https://www.guardsquare.com/en/proguard>`_ and Corda's ``JarFilter`` Gradle
+  plugin. Corda developers configure these tools by applying Corda's ``@Deterministic`` and ``@NonDeterministic``
+  annotations to elements of ``core`` and ``serialization`` as described `here <deterministic_annotations_>`_.
```
Comment (mikehearn) [inline docs/source/deterministic-modules.rst]: need double underscore for links with ambiguous text like "here" (see my email to corda-dev on the topic of rst markup errors).

Comment (chrisr3) [reply docs/source/deterministic-modules.rst]: This is the link target for the "here" text:
```
.. _deterministic_annotations:

```
This shouldn't be ambiguous.

File: docs/source/deterministic-modules.rst [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+
+    ProGuard works by calculating how much code is reachable from given "entry points", and in our case these entry
+    points are the ``@Deterministic`` classes. The unreachable classes are then discarded by ProGuard's ``shrink``
+    option.
+ #. The remaining classes may still contain non-deterministic code. However, there is no way of writing a ProGuard rule
+    explicitly to discard anything. Consider the following class:
+
+    .. sourcecode:: kotlin
+
+        @CordaSerializable
```
Comment (exFalso) [inline docs/source/deterministic-modules.rst]: Perhaps it's better to link source code using `literalinclude` so we don't have to keep this piece of code updated with `UniqueIdentifier`

File: docs/source/deterministic-modules.rst [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+    any more code that has now become unreachable.
+ #. The remaining classes define our deterministic subset. However, the ``@kotlin.Metadata`` annotations on the compiled
+    Kotlin classes still contain references to all of the functions and properties that ProGuard has deleted. Therefore
+    we now use the ``JarFilter`` to delete these references, as otherwise the Kotlin compiler will pretend that the
+    deleted functions and properties are still present.
+ #. Finally, we use ProGuard again to validate our JAR against the deterministic ``rt.jar``:
+
+    .. sourcecode:: groovy
+
+        task checkDeterminism(type: ProGuardTask, dependsOn: jdkTask) {
```
Comment (exFalso) [inline docs/source/deterministic-modules.rst]: Similarly, better to use `literalinclude`. See hello-world-state.rst for an example

File: docs/source/deterministic-modules.rst (hunk 2 above)
Comment (chrisr3) [reply docs/source/deterministic-modules.rst]: I can do that, but suspect I'll need to get _both_ #3262 _and_ #3297 merged first.

Comment (chrisr3) [reply docs/source/deterministic-modules.rst]: Although to be fair, if someone were to change `UniqueIdentifer` _not_ to have these secondary constructors then they'd break my document instead :wink:!

I don't think that this _particular_ example benefits from being linked to the actual code.

**Pull request merged**